### PR TITLE
Usagov 1856 datalayer twig

### PR DESCRIPTION
**Pull request by omerida**
## Jira Task

https://cm-jira.usa.gov/browse/USAGOV-1856

## Description

Moves building the datalayer items out of TWIG templates into the preprocess hook used to gather the node related information. Main change is directly getting the breadcrumb items the block accessed but before any of the items were rendered. Replicates the same logic that was in the twig layer in a DatalayerBuilder object.

## Type of Changes

- [ ] New Feature
- [ ] Bugfix
- [ ] Frontend (Twig, Sass, JS)
  - Add screenshot showing what it should look like
- [ ] Drupal Config (requires "drush cim")
- [ ] New Modules (requires rebuild)
- [ ] Documentation
- [ ] Infrastructure
  - [ ] CMS
  - [ ] WAF
  - [ ] Egress
  - [ ] Tools
- [x] Other 
    -  Refactoring which shouldn't have any side effects.

## Testing Instructions

See documentation in Datalayer_for_GTM.md for how to run the verify script to check that there are no changes in the output.

### Change Requirements

- [ ] Requires New Documentation (Link: {})
- [ ] Requires New Config
- [ ] Requires New Content

### Validation Steps

- Verify script should flag that there are no unexpected changes in the datalayer across all exported pages.

## Security Review

- [ ] Adds/updates software (including a library or Drupal module)
- [ ] Communication with external service
- [ ] Changes permissions or workflow
- [ ] Requires SSPP updates

## Reviewer Reminders

- Reviewed code changes
- Reviewed functionality
- Security review complete or not required

## Post PR Approval Instructions

Follow these steps as soon as you merge the new changes.

1. Go to the [USAGov Circle CI project](https://app.circleci.com/pipelines/github/usagov/usagov-2021).
2. Find the commit of this pull request.
3. Build and deploy the changes.
4. Update the Jira ticket by changing the ticket status to `Review in Test` and add a comment. State whether the change is already visible on [cms-dev.usa.gov](http://cms-dev.usa.gov/) and [beta-dev.usa.gov](http://beta-dev.usa.gov/), or if the deployment is still in process.

### CODE REVIEW

File: web/modules/custom/usa_twig_vars/usa_twig_vars.module [hunk 1]
```diff
@@ -46,82 +44,53 @@ function usa_twig_vars_preprocess(&$variables, $hook) {
...
+          $usa_twig_vars['datalayer'] = $datalayer->build();
           unset($isFront);
+
+          // Cleanup vars.
+          // Promote variables figured out via datalayer builder that
+          // we initialized earlier.
+          // Need to get them from the array return after build() above.
+          $usa_twig_vars['id'] = $node->id();
+          $usa_twig_vars['nodetype'] = $usa_twig_vars['datalayer']['contentType'];
+          $usa_twig_vars['pagetype'] = $usa_twig_vars['datalayer']['basicPageSubtype'];
```
Comment (akf) [inline web/modules/custom/usa_twig_vars/usa_twig_vars.module]: I'm seeing a warning in my error log that 'basicPageSubtype' is an undefined array key. It's 'basicPagesubtype' on line 78.

Comment (akf) [reply web/modules/custom/usa_twig_vars/usa_twig_vars.module]: I would expect this to have a lot more effects than it does! page.html.twig has a use of it that affects the spacing under the breadcrumbs for "Navigation Cards Page" and I also see it in block--menu-block--navigation-page-items.html.twig but visually I don't see the effect. 

Anyway, just fix it so twig still has "pagetype" defined and all will be well.

**Review by akf — CHANGES_REQUESTED** [review]
Mostly looking good. I caught one issue, and only because I had my logs scrolling while running the verify-pagelayer-data.php script.

File: web/modules/custom/usa_twig_vars/usa_twig_vars.module (hunk 1 above)
Comment (omerida) [reply web/modules/custom/usa_twig_vars/usa_twig_vars.module]: I've fixed the typo and then realized I didn't have Drupal configured to show errors and warning on screen after my last DB sync. Grrr.

In theory, at the higher levels, this is something PHPStan could catch once you describe array shapes.

**Review by akf — COMMENTED** [review]
I realized today that this PR was still outstanding and I tried to give it a good review. I have run across another very puzzling thing. I'm going to add my comments on the Jira ticket because I'm going to link to some files I've stored on GDrive. 

The short description is, as far as I can tell, the HTML output is perfect (or, at least as good as what we're getting on the dev branch). But I did a comparison of the published_pages_report CSV and it found some anomalies.

**Review by akf — CHANGES_REQUESTED** [review]
I found an edge case (empirically): Pages that are seven or more levels deep in the taxonomy are now getting that information included in the "taxonomy-data" script, and it's bleeding through to the published_pages.csv. 

Examples are the final pages of many of the "wizards" currently on usa.gov. As one example: https://www.usa.gov/where-report-scams/what-type-scam-do-you-need-report/housing-scam/moving-scam/moving-scam-within-state

On prod, we see this: 

```
<script id="taxonomy-data">
    dataLayer = [{
    "nodeID": "1895",
    "contentType": "wizard_step",
    "language": "en",
    "homepageTest": "not_homepage",
    "basicPagesubType": null,
    "Page_Type": "wizard_step",
    "Taxonomy_Text_1": "Home",
    "Taxonomy_Text_2": "Scams and Fraud",
    "Taxonomy_Text_3": "Where to report a scam",
    "Taxonomy_Text_4": "What type of scam do you need to report?",
    "Taxonomy_Text_5": "Housing scam",
    "Taxonomy_Text_6": "Moving scam",
    "Taxonomy_URL_1": "/",
    "Taxonomy_URL_2": "/scams-and-fraud",
    "Taxonomy_URL_3": "/where-report-scams",
    "Taxonomy_URL_4": "/where-report-scams/what-type-scam-do-you-need-report",
    "Taxonomy_URL_5": "/where-report-scams/what-type-scam-do-you-need-report/housing-scam",
    "Taxonomy_URL_6": "/where-report-scams/what-type-scam-do-you-need-report/housing-scam/moving-scam",
    "hasBenefitCategory": false
}];
  </script>
```

On this branch, I get this: 

```
<script id="taxonomy-data">
    dataLayer = [{
    "nodeID": "1895",
    "language": "en",
    "homepageTest": "not_homepage",
    "basicPagesubType": null,
    "contentType": "wizard_step",
    "Page_Type": "wizard_step",
    "Taxonomy_Text_1": "Home",
    "Taxonomy_Text_2": "Scams and Fraud",
    "Taxonomy_Text_3": "Where to report a scam",
    "Taxonomy_Text_4": "What type of scam do you need to report?",
    "Taxonomy_Text_5": "Housing scam",
    "Taxonomy_Text_6": "Moving scam",
    "Taxonomy_Text_7": "Moving scam (within the same state)",
    "Taxonomy_URL_1": "/",
    "Taxonomy_URL_2": "/scams-and-fraud",
    "Taxonomy_URL_3": "/where-report-scams",
    "Taxonomy_URL_4": "/where-report-scams/what-type-scam-do-you-need-report",
    "Taxonomy_URL_5": "/where-report-scams/what-type-scam-do-you-need-report/housing-scam",
    "Taxonomy_URL_6": "/where-report-scams/what-type-scam-do-you-need-report/housing-scam/moving-scam",
    "Taxonomy_URL_7": "/where-report-scams/what-type-scam-do-you-need-report/housing-scam/moving-scam/moving-scam-within-state",
    "hasBenefitCategory": false
}];
  </script>
```

That's logical, but it's not what the stakeholder wanted. In the published pages report, these additional values are populating the last cells of the row for the page, which are currently the "hasBenefitCategory" and (unlabeled column that lists categories; do we mean to have this at all?) 

That is, happily, the only issue I can see!

**Comment by omerida** [comment]
Good find. I thought the wizard pages had their own template building the
taxonomy url paths but I'll take a look to see why it's affecting the
published pages report,

**Comment by akf** [comment]
I should probably clarify (as you might already have seen by now) -- the
page I cited is an "old" wizard page, not one of the new taxonomy-based
wizards. The latter use their own template. The former are a distinct page
type, but are probably more like basic pages.

**Comment by omerida** [comment]
I fixed the issue where its creating all the breadcrumbs regardless of depth. I ran the verify script (modified to detect when there are more than 6 levels too) and it found no differences. I also checked the published pages csv generated by this branch against one generated by dev and they were the same.

**Comment by omerida** [comment]
I've merged and fixed conflicts from linting. I ran tome agains dev and this branch locally. HTML output was the same except for the position of one of the datalayer attributes changing because of moving more things out of twig in this branch. The output of the published pages CSV file were identical for me. I'm hoping merging this in might fix/prevent some of the weirdness I run into, not predictably, when trying to get tome to work with more than one path per process.

**Comment by akf** [comment]
Re-review started! I'm kicking this off by running tome on the dev branch, so I'll probably complete the review tomorrow.

**Review by akf — APPROVED** [review]
Looks good; I didn't find any changes in output compared to the current dev branch (my investigation was not totally comprehensive).

**Pull request merged**